### PR TITLE
[Enhancement] Improve user interface for Manual Assessment of Programming Exercises

### PR DESCRIPTION
**Pull request by fde312**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

We want to improve the UI of the manual assessment of programming exercises and align it more to the other exercises.

### Description

Comment from @krusche in #2223 : 
- [x] Only show 1 result (top right corner) at the same time based on the last automatic one, that automatically adapts to the manual assessment input. This means it should also be clickable and show the current manual and automatic feedback in the popup
- [x] Remove the Grading Instructions below the code editor
- [x]  Move the button Download repos to the upper right corner left to the result, rename it to "Download repo" on this page, because only one single repository will be downloaded
- [x]  Also move the link "Go to repository" to the upper right corner left to the result
- [x] The result should show the current score in percentage (in addition it can show the score in points in case this is helpful, but then both needs to be in sync)
- [ ]  In the exercise dashboard, change the buttons "New Result" and "Update result" to how they are named in other exercise types. This has to be consistent. -> Follow-up PR

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create programming exercise with SCA and without SCA
4. Participate, Assess, Complain and Assess the Complain (everything should work as intended, try out different scenarios)
5. Try also out the Configure Grading see PR test steps of #2223 (steps for programming exercise)

### Screenshots

Screenshots are made with a SCA Programming Exercise with 100 maxScore and 10 bonus points, and deduction for code style issues each 1 point.

**NOTE**: Do not wonder why the issues are not listed in the result string before the configure grading test case change. I used submissions, where adding the issue numbers into the result string, was not yet implemented.

New layout of assessment view:
image
image

When changing the feedback elements, result string adapts automatically (did not press on save or submit)
image

Also displayed in the dialog:
image

After clicking on save the preliminary assessment is visible. (still not visible for students)
image

After submitting (tutor view):
image

Student sees the new result string:
image

After Changing configure grading (1 extra point for correct bubbleSortTest), the result string adapts accordingly:
NOTE: Now the issues are listed in the result string as well (were not before because of 'old' submission)
image
image

After student complaint and assessing the complaint (adding 10 points in inline feedback and 10 points unreferenced)
image

image

image

Student view after assessing the complaint:
image
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 1]
```diff
@@ -109,8 +109,22 @@ public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exe
 
         // Update the result that was complained about with the new feedback
         originalResult.updateAllFeedbackItems(assessmentUpdate.getFeedbacks());
-        Double calculatedScore = calculateTotalScore(originalResult.getFeedbacks());
-        return submitResult(originalResult, exercise, calculatedScore);
+        if (exercise instanceof ProgrammingExercise) {
+            double points = ((ProgrammingAssessmentService) this).calculateTotalScore(originalResult);
```
Comment (kloessst) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:113]: Do we need the exercise type case distinction here? If `ProgrammingAssessmentService` inherits from `AssessmentService`, you should override `updateAssessmentAfterComplaint` and call it on the child class.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 2]
```diff
@@ -144,7 +144,19 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+        if (newResult.getResultString() == null) {
+            throw new BadRequestAlertException("Result string is required.", ENTITY_NAME, "resultStringNull");
+        }
+        else if (newResult.getResultString().length() > 255) {
+            throw new BadRequestAlertException("Result string is too long.", ENTITY_NAME, "resultStringNull");
+        }
+        else if (newResult.getScore() == null) {
+            throw new BadRequestAlertException("Score is required.", ENTITY_NAME, "scoreNull");
+        }
+        else if (newResult.getScore() < 100 && newResult.isSuccessful()) {
```
Comment (kloessst) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:159]: I'd suggest moving the validation of `newResult` as far up as possible in the method. Validation should be done before services are used.

**Review by kloessst — CHANGES_REQUESTED** [review]
Everything worked fine in the testing session. Good job! Just a few remarks about the Java code:

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java (hunk 2 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:159]: The calls which are done before are also checks, e.g. if the user has enough roles etc. I would leave it like it is, as I think it is equally important to check if he has the rights to submit/save a result or that the result has the correct format. :)

File: src/main/java/de/tum/in/www1/artemis/domain/Result.java [hunk 3]
```diff
@@ -137,6 +137,24 @@ public void setResultString(Double totalScore, @Nullable Double maxScore) {
...
     }
 
+    /**
+     * Builds a part of the result string of manual results for programming exercises
+     *
+     * @param totalScore total amount of scored points
+     * @param maxScore   maximum score reachable at corresponding exercise
+     * @return String with result string for manual result of a progrmaming exercise
+     */
+    @JsonIgnore
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/domain/Result.java:147]: Fixed in `5a4957277 `

File: src/main/java/de/tum/in/www1/artemis/domain/Result.java [hunk 4]
```diff
@@ -137,6 +137,24 @@ public void setResultString(Double totalScore, @Nullable Double maxScore) {
...
+    public String createResultStringForManualResult(Double totalScore, @Nullable Double maxScore) {
+        DecimalFormat formatter = new DecimalFormat("#.##");
+        if (maxScore == null) {
+            return formatter.format(totalScore) + " points";
+        }
+        else {
+            return formatter.format(totalScore) + " of " + formatter.format(maxScore) + " points";
+        }
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/domain/Result.java:155]: Good catch! Thanks! Fixed in `5a4957277 `

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java (hunk 1 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:113]: I cannot override the method as there is no `calculateTotalScore(Result result)` in `AssessmentService` only for programming exercises we have a special way on calculating the points the student has :)

Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:113]: I understand your point, but it is basically then code duplication except of the last part. I am not quite sure it this is then better maintainable?

**Review by kloessst — APPROVED** [review]
Approve after @fde312 implemented the suggestions.

**Review by anditurdiu — APPROVED** [review]
Works well! The improvements are really cool

**Review by stefanwaldhauser — APPROVED** [review]
Tested it without problems in the testing session

**Review by schultek — APPROVED** [review]
Everything works and looks good. 👍

**Review by MaisiKoleni — APPROVED** [review]
Looks good to me and works well on TS1.

**Pull request merged**